src/Transform/ONNX/ShapeInferencePass.cpp

```diff
@@ -164,10 +152,14 @@ class ShapeInferencePass : public mlir::PassWrapper<ShapeInferencePass,
   /*!
    *  Check if the given operation has a dynamically shaped result.
    */
-  static bool returnsDynamicShape(Operation *op) {
+  static bool returnsDynamicOrUnknownShape(Operation *op) {
     return llvm::any_of(op->getResultTypes(), [](Type result_type) {
-      return !result_type.isa<NoneType>() &&
-             !result_type.isa<RankedTensorType>();
+      if (result_type.isa<RankedTensorType>()) {
+        return llvm::any_of(result_type.dyn_cast<RankedTensorType>().getShape(),
+            [](int64_t dim) { return dim < 0; });
+      } else {
+        return !result_type.isa<NoneType>();
```

Comment:
Wouldn't it be better to explicitly test if this is an unranked tensor-type? Is there an assumption right now that the only types of values are either tensors or NoneType? I was wondering what happens in the future, when there is a mix of code returning primitive types, or other types.

Reply:
@gramalingam I read your comment after I merged the PR. Yes, it is better to check unranked tensor-type.